pkg/reconciler/broker/broker.go

```diff
@@ -193,6 +173,29 @@ func (r *Reconciler) reconcileBroker(ctx context.Context, b *brokerv1beta1.Broke
 	}
 	b.Status.PropagateIngressAvailability(ingressEndpoints)
 
+	// Filter by `eventing.knative.dev/broker: <name>` here
+	// to get only the triggers for this broker. The trigger webhook will
+	// ensure that triggers are always labeled with their broker name.
+	triggers, err := r.triggerLister.Triggers(b.Namespace).List(labels.SelectorFromSet(map[string]string{eventing.BrokerLabelKey: b.Name}))
+	if err != nil {
+		logger.Error("Problem listing triggers", zap.Error(err))
+		return err
+	}
+
+	r.reconcileConfig(ctx, b, triggers)
+	// Update config map
+	r.flagTargetsForUpdate()
+	return nil
+}
+
+// reconcileConfig reconstructs the data entry for the given broker in targets-config.
+func (r *Reconciler) reconcileConfig(ctx context.Context, b *brokerv1beta1.Broker, triggers []*brokerv1beta1.Trigger) {
+	// TODO Maybe get rid of BrokerMutation and add Delete() and Upsert(broker) methods to TargetsConfig. Now we always
+	//  delete or update the entire broker entry and we don't need partial updates per trigger.
+	// The code can be simplified to r.targetsConfig.Upsert(brokerConfigEntry)
+	// First delete the broker entry.
+	r.targetsConfig.MutateBroker(b.Namespace, b.Name, func(m config.BrokerMutation) { m.Delete() })
+	// Then reconstruct the broker entry and insert it
 	r.targetsConfig.MutateBroker(b.Namespace, b.Name, func(m config.BrokerMutation) {
```

Comment:
This should be done in the same mutation as the triggers below.

Reply:
Actually "delete" cannot be done in the same mutation (in the existing implementation) as it will ignore all other operations. This is because the mutation stages all operations without order (my bad) before "committing". If a "delete" operation appears in the mutation, it will treat it as "you want to change broker data and delete it in the same transaction" which essentially equals to "delete the broker". In my other comment, I proposed to add a "reset" function to the mutation.

Reply:
Made a fix: https://github.com/google/knative-gcp/pull/1142

Reply:
It seems to me making `Delete` a separate Mutation is good for now. We can revisit the interface as a follow up.
I am not sure a `reset` is needed. I believe a single `Upsert` is enough. With regards to the map in the config.Broker, I don't think that's a big deal. If we really want to hide that, we can add an interface like so: `Upsert(broker, triggers)` and let the config handle the mapping.

Reply:
Yea, I didn't add a "reset" in the fix (I fixed "delete" instead). Do notice that you have to stage triggers in a var before you upsert. The actual amount of code would likely look similar to what you have today.

Reply:
Yeah the amount of code is likely similar. But that's not the point. I was trying to achieve: 1, simplified interface as the interaction is simplified; 2. avoid maintaining states and concurrency in the config lib since we don't need that with brokercell.

Anyways, we don't need to figure out the exact change now. We can do that when we reconcile the config in the brokercell.